Lean: Adding features to support structs and bitfields

lfrenot opened the pull request:
This implements the translation of bitfields, and of the operations on structs (accessing a field, editing a field, making a new instance of a struct)

lfrenot committed:
First batch of features

lfrenot committed:
Adding features

lfrenot committed:
formatting

lfrenot committed:
update_subrange implementation

lfrenot committed:
Cleanup registers

lfrenot committed:
updated test

lfrenot committed:
formatting

lfrenot committed:
fix test name

lfrenot committed:
fix expected output

javra commented:
Looks good! The convention for Lean files is to start with an upper case, I think we'd better stick to that

lfrenot committed:
Capitalizing the Lean files

lfrenot committed:
Larger struct test file

lfrenot committed:
Typo fixes

lfrenot committed:
missing edit for the rebase

lfrenot committed:
Removing useless changes

lfrenot committed:
Removing the addition import files

because the imports in Sail.lean needed the output folder's name to work properly

alexkeizer commented on src/sail_lean_backend/Sail/Sail.lean:
```diff
@@ -15,5 +15,16 @@ def truncate {w: Nat} (x: BitVec w) (w': Nat) : BitVec w' :=
 def truncateLSB {w: Nat} (x: BitVec w) (w': Nat) : BitVec w' :=
   x.extractLsb' 0 w'
 
+def extractLsb {w: Nat} (x: BitVec w) (hi lo: Nat) : BitVec (hi - lo + 1) :=
+  x.extractLsb hi lo
```
```suggestion
def extractLsb {w : Nat} (x : BitVec w) (hi lo : Nat) : BitVec (hi - lo + 1) :=
  x.extractLsb hi lo
```

Lean style puts a space before the `:` in arguments. I did notice the rest of the file has the same formatting, so we might want to address this in a separate formatting PR and fix the whole file in one go.

alexkeizer commented on src/sail_lean_backend/Sail/Sail.lean:
```diff
@@ -15,5 +15,16 @@ def truncate {w: Nat} (x: BitVec w) (w': Nat) : BitVec w' :=
 def truncateLSB {w: Nat} (x: BitVec w) (w': Nat) : BitVec w' :=
   x.extractLsb' 0 w'
```
```suggestion
def truncateLsb {w: Nat} (x: BitVec w) (w': Nat) : BitVec w' :=
  x.extractLsb' 0 w'
```

We generally treat `Lsb` as one word, for the sake of capitalization. (Outside of the diff of this PR, so feel free to ignore)

alexkeizer commented on src/sail_lean_backend/Sail/Sail.lean:
```diff
@@ -15,5 +15,16 @@ def truncate {w: Nat} (x: BitVec w) (w': Nat) : BitVec w' :=
+def extractLsb {w: Nat} (x: BitVec w) (hi lo: Nat) : BitVec (hi - lo + 1) :=
+  x.extractLsb hi lo
+
+def update_subrange' {w: Nat} (x: BitVec w) (start len: Nat) (y: BitVec len) : BitVec w :=
+  let mask := ~~~(((BitVec.allOnes len).zeroExtend w) <<< start)
+  let y' := mask ||| ((y.zeroExtend w) <<< start)
+  x &&& y'
+
+def update_subrange {w: Nat} (x: BitVec w) (hi lo: Nat) (y: BitVec (hi - lo + 1)) : BitVec w :=
+  update_subrange' x lo _ y
```
```suggestion
def updateSubrange' {w: Nat} (x: BitVec w) (start len: Nat) (y: BitVec len) : BitVec w :=
  let mask := ~~~(((BitVec.allOnes len).zeroExtend w) <<< start)
  let y' := mask ||| ((y.zeroExtend w) <<< start)
  x &&& y'

def updateSubrange {w: Nat} (x: BitVec w) (hi lo: Nat) (y: BitVec (hi - lo + 1)) : BitVec w :=
  updateSubrange' x lo _ y
```

Lean style uses `lowerCamelCase` for definitions

lfrenot committed:
Adding back registerRef

lfrenot committed:
Formatting

lfrenot committed:
Fixing Lean style in Sail.lean

alexkeizer commented on src/sail_lean_backend/Sail/Sail.lean:
```diff
@@ -1,19 +1,37 @@
 end BitVec
 end Sail
+
+structure registerRef (regstate regval a : Type) where
```
```suggestion
structure RegisterRef (regstate regval a : Type) where
```

In this case, `RegisterRef` is a type(-valued function) and thus is written `UpperCamelCase`! (Sorry for not catching that before)

alexkeizer commented on src/sail_lean_backend/Sail/Sail.lean:
```diff
@@ -1,19 +1,37 @@
 
-def truncate {w: Nat} (x: BitVec w) (w': Nat) : BitVec w' :=
+def truncate {w : Nat} (x : BitVec w) (w' : Nat) : BitVec w' :=
   x.truncate w'
 
-def truncateLSB {w: Nat} (x: BitVec w) (w': Nat) : BitVec w' :=
+def truncateLsb {w : Nat} (x : BitVec w) (w' : Nat) : BitVec w' :=
   x.extractLsb' 0 w'
```
Again, this is stuff outside the scope of your PR diff, but since you're cleaning it up I'll comment anyway:
`truncate` and `truncateLsb` are [provably equivalent](https://live.lean-lang.org/#codez=CYUwZgBALgTgrgOwMYEMoggbwO4QFwQByaAvhABQAe+EAQgJZQBqISE2AlBdgOQ3FQuBBs1bs+eALwAoCBEoA6WIlTpx06aEjLkaEABkAzgCMsuAgLJUaIlm07cJRNELqM74/DLmKQlWChIUEbGfAAM6tJ+KAC2AA4ANhgEAAI6qhi4ktDwuughXhDGAJ6yEGCIflDy6nKG9PEQANrpegA0OSp6IR22rApVMIHBJjwAumWTQA). It's unclear to me why we need both definitions

javra replied:
Yes, let's delete `truncateLsb` and avoid redundant defs.

Thank you for the annotations, @alexkeizer!

lfrenot replied:
Oh I see
This is my bad, `truncateLsb` is supposed to extract the most significant bits, the first argument is wrong here.
Since this changes more, I think I'll just open a separate PR to fix this and the formatting

lfrenot committed:
changing registerRef to UpperCamelCase

lfrenot committed:
Fixing after rebase

lfrenot commented:
This will fail until https://github.com/rems-project/sail/pull/870 is merged, because I need the parentheses fix it introduces

lfrenot committed:
Adding a mk_struct test

lfrenot committed:
updated test output for rebase

javra commented on test/lean/struct.expected.lean:
```diff
@@ -4,11 +4,21 @@ open Sail
+
+/-- Type quantifiers: i : Int -/
+def mk_struct (i : Int) (b : (BitVec 1)) : My_struct :=
+  {field1 := i, field2 := b}
```
Canonical Lean style would mandate spaces inside the curly brackets. But we could also print this as

```lean
def mk_struct (i : Int) (b : (BitVec 1)) : My_struct :=
  { field1 := i
    field2 := b }
```
which might be better for bigger terms

lfrenot replied:
I agree with your solution, I'll do that

javra replied:
(Note that we don't even need the `,` if we do the indentation right)

lfrenot replied:
To be sure I follow the style: should it be
```
def mk_struct (i : Int) (b : (BitVec 1)) : My_struct :=
  {
    field1 := i
    field2 := b
  }
```
or
```
def mk_struct (i : Int) (b : (BitVec 1)) : My_struct :=
  { field1 := i
    field2 := b }

javra replied:
The latter

lfrenot replied:
Thanks

lfrenot replied:
Done

alexkeizer replied:
Note that the following should also be a valid spelling of the same definition:
```lean
def mk_struct (i : Int) (b : (BitVec 1)) : My_struct where
  field1 := i
  field2 := b
```

Not that there's any clear benefit over this style, so I'm not suggesting we switch to it necessarily, just wanted to make you aware of the full space of options

lfrenot replied:
This wouldn't be practical here I think, since the `:=` is in the function definition, not in the expression.
but good to know

javra replied:
Yes I think keeping the `:=` is easier. This might be a task that could be given to the mighty future Lean reformatter ™

lfrenot committed:
Lean style update
